src/Provider/Provider.Web/Middleware/ProviderAccountHandler.cs

```diff
@@ -81,16 +86,63 @@ private async Task<bool> HasUkprnAuthorizationAsync(AuthorizationHandlerContext
 
             return false;
         }
-        
-        private async Task EnsureProviderIsSetup(HttpContext context, long ukprn)
+
+        private async Task<bool> HasRoatpAuthorizationAsync(AuthorizationHandlerContext context)
         {
-            var key = string.Format(CookieNames.SetupProvider, ukprn);
+            if (HasDoneOncePerAuthorizedSessionActions(context))
+                return true;
 
-            if (context.Request.Cookies[key] == null)
+            try
             {
-                await _client.SetupProviderAsync(ukprn);
-                context.Response.Cookies.Append(key, "1", EsfaCookieOptions.GetDefaultHttpCookieOption(_hostingEnvironment));
+                var ukprnFromClaim = context.User.FindFirst(_ukprnClaimFinderPredicate).Value;
+                var allProviders = await _vacancyClient.GetAllTrainingProvidersAsync();
```

Comment:
Just a thought, haven't tried it myself but could we not throw an Authorisation exception from the ProviderAccountHandler which indicates the roatp failure instead of making potential network calls from this ErrorController.

Reply:
Unfortunately there is no way to add context to the 403 that the `AuthorizationHandler` raises so there is no way to find out why the 403 occurred from the error controller.  We could `RedirectToRoute` directly from the handler to a known route but that would mean introducing a special case for this particular exception which would add complexity.  At the moment all 403 errors go through the error controller which I think is a good thing.  Its not ideal however I think overall its probably the best we can do.